open-craft/opencraft
# feat: Display and trigger deployments

## Description

This PR implements deployment method indication for instances to know if the instance is deployed by Grove or deployed on OpenStack VMs.

Also, the PR adds the possibility of triggering new deployments -- which creates a new merge commit as of now by Grove, which will be changed later on.

## Supporting information

This PR heavily depends on a working Grove cluster.

### Screenshots

_Ocim_

<img width="1276" alt="Screenshot 2022-05-19 at 14 05 41" src="https://user-images.githubusercontent.com/19173947/169289414-f454d02a-f814-43a3-8cdd-8f45a9a0c3f2.png">

_Grove_

<img width="1409" alt="Screenshot 2022-05-19 at 13 58 24" src="https://user-images.githubusercontent.com/19173947/169288209-b433f455-3650-47a0-8ba4-81b863f00582.png">

## Testing instructions

__Testing requires local Ocim installation this time!__

1. Checkout the branch
2. Update `.env` with the Grove cluster configuration details
3. Run migrations
4. Start the backend and frontend
5. Start `ngrok` or a similar tool to expose the instance
6. Find the cluster settings in GitLab and navigate to "Webhook settings"
7. Add a new webhook with (only) "Pipeline events" checked for `<URL>/api/webhook/`
8. Register a new instance (through the normal registration flow)
9. Check that the instance is created in Ocim
10. Check that the deployment appears for the instance
11. Check that the pipeline is triggered

__Testing on stage__

1. Register a new instance (through the normal registration flow)
2. Check that the instance is created in Ocim
3. Check that the deployment appears for the instance
4. Check that the pipeline is triggered

## Deadline

ASAP

## Other information

1. During the implementation, I found many smaller issues which should be investigated, adjusted, or reworked. The PR has no change set for these as it would be way out of scope. Follow-up tickets will be created for SE.273.
13. Grove <> Ocim integration is not complete yet, though it will be resolved in SE.273 (in conjunction with the removal of the OpenStack deployment method).
